### PR TITLE
2.0 preview

### PR DESCRIPTION
**Pull request by e111077**
DO NOT MERGE

notable changes
- created `Polymer.PlatinumSwRegisterableBehavior` which should be used by all elements that should register within a `<platinum-sw-register>`
- `<platinum-sw-register>` no longer iterates through the light dom when attached. platinum-sw-registerable elements can now be attached after first load as it is now event-driven registration

### CODE REVIEW

**Comment by jeffposnick** [comment]
CC: @tjsavage @addyosmani 

Sorry to drop in after a bunch of work has already been completed. As the original author of the Polymer 1.x version of this library, I'm of the opinion that deprecation would be more appropriate than porting to Polymer 2.x.

`polymer-cli` ships with more comprehensive service worker support via `sw-precache`, and developers are much more likely to end up with a web app that follows the best practices we are promoting if they went with that as a starting point.

**Comment by e111077** [comment]
@jeffposnick We have already made the decision to deprecate this element, but we found it necessary to migrate all elements despite deprecation to at least the hybrid 2.0 mode. We will lock the bower dep to `~3.4.0`. Thanks for the heads up

File: platinum-sw-register.html [hunk 1]
```diff
@@ -108,7 +108,8 @@
         // Fallback to './' as a default, though current browsers that don't support
         // document.currentScript also don't support service workers.
         value: document._currentScript ? document._currentScript.baseURI :
-          (document.currentScript ? document.currentScript.baseURI : './')
+          (HTMLImports.importForElement ? HTMLImports.importForElement(document.currentScript).baseURI :
```
Comment (e111077) [inline platinum-sw-register.html]: This has to be a one-liner or else the `currentScript` is different.

Comment (notwaldorf) [inline platinum-sw-register.html]: :shipit:

File: platinum-sw-registerable-behavior.html [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  /**
+   * `Polymer.PlatinumSwRegisterableBehavior` fires un/registration events which
+   * are handled by `platinum-sw-register`. Also provides abstract methods for
+   * each platinum-sw registerable element.
+   * 
+   * @polymerBehavior Polymer.PlatinumSwRegisterableBehavior
+   */
+  Polymer.PlatinumSwRegisterableBehavior = {
+    properties: {
+      /**
```
Comment (notwaldorf) [inline platinum-sw-registerable-behavior.html]: Prolly don't need this properties block

**Comment by e111077** [comment]
I changed it to a mutation observer, but tests inexplicably now fail only on Chrome.

It's odd because debuggers show that the registration does the exact same stuff

**Comment by e111077** [comment]
@notwaldorf AH I remember why I used this implementation. The issue is that in 1.x register needs to be only called once and after all of the children in the light dom to be polymer elements. Currently, with a mutation observer, the children in the light dom return false for `child instanceof Polymer.Element`

File: platinum-sw-registerable-behavior.html (hunk 2 above)
Comment (e111077) [inline platinum-sw-registerable-behavior.html]: Gone!

**Comment by e111077** [comment]
@notwaldorf 👍 👍 ❓ 👀

**Comment by notwaldorf** [comment]
@e111077 but but but "// Note: requires all children to have `platinum-sw-registerable-behavior`" is a gigantor breaking change no?

File: platinum-sw-register.html [hunk 3]
```diff
@@ -376,8 +374,26 @@
     },
 
     attached: function() {
+      var isSupported = 'serviceWorker' in navigator;
+
+      if (!isSupported) {
```
Comment (notwaldorf) [inline platinum-sw-register.html]: nice!

File: platinum-sw-register.html [hunk 4]
```diff
@@ -376,8 +374,26 @@
...
+        this._setState('unsupported');
+        return;
+      }
+
       if (this.autoRegister) {
-        this.async(this.register);
+        if (document.readyState !== 'complete' &&
+            document.readyState !== 'interactive') {
+
+          document.onreadystatechange = function() {
```
Comment (notwaldorf) [inline platinum-sw-register.html]: This is the 2.0 change, right? Wanna add a small comment as to why it's needed? ("need to figure out if children are upgraded" or some sorts)

Comment (e111077) [inline platinum-sw-register.html]: @sorvell claims that the children in the light dom should be upgraded after `document.readyState === 'interactive'` plus a rAF assuming all the light dom webcomponents are html-imported synchronously.

The polymer 1 version of this assumed that they were all already upgraded and would simply go through them and see if they had a function, called it, and then shoved the concatenated results into a SW.

Polymer 2 upgrade / lifecycle order doesn't work in the same way

**Pull request merged**